File: src/main/java/gargoyle/ct/VirtualResource.java
package gargoyle.ct;

import java.io.IOException;
import java.net.URL;
import java.util.Locale;
import java.util.Objects;
import java.util.ResourceBundle.Control;

public class VirtualResource extends AbstractResource {

    private final Resource baseResource;

    private Locale locale;

    protected VirtualResource(Resource baseResource, String location) {
        super(location);
        this.baseResource = baseResource;
    }

    protected VirtualResource(String location) {
        super(location);
        baseResource = this;
    }

    @Override
    protected VirtualResource createResource(Resource base, String location) {
        return base == null ? new VirtualResource(location) : new VirtualResource(base, location);
    }

    @Override
    public Resource forLocale(Locale locale) {
        Control ctrl = Control.getControl(Control.FORMAT_DEFAULT);
        String location = baseResource.getLocation();
        int i = location.lastIndexOf('.');
        String baseName = location.substring(0, i);
        String suffix = location.substring(i + 1);
        for (Locale specificLocale : ctrl.getCandidateLocales(baseName, locale)) {
            String loc = ctrl.toResourceName(ctrl.toBundleName(baseName, specificLocale), suffix);
            VirtualResource resource = createResource(baseResource, loc);
            if (resource.exists()) {
                resource.locale = locale;
                return resource;
            }
        }
        return null;
    }

    @Override
    public Locale getLocale() {
        return locale;
    }

    public String getRelativeLocation() {
        String thisLocation = getLocation();
        String baseLocation = baseResource.getLocation();
        int minLength = Math.min(thisLocation.length(), baseLocation.length());
        for (int i = 0; i < minLength; i++) {
            if (thisLocation.charAt(i) != baseLocation.charAt(i)) {
                return thisLocation.substring(i);
            }
        }
        return thisLocation;
    }

    @Override
    public URL toURL() throws IOException {
        return Objects.equals(baseResource, this) ? super.toURL() : new URL(baseResource.toURL(), getLocation());
    }

    @Override
    public Resource withExtension(String extension) {
        String location = getLocation();
        String loc = location.substring(0, location.lastIndexOf('.')) + "." + extension;
        return createResource(null, loc);
    }
}

File: util/src/main/kotlin/gargoyle/ct/util/resource/internal/VirtualResource.kt
package gargoyle.ct.util.resource.internal

import gargoyle.ct.util.resource.Resource
import java.io.File
import java.net.URL
import java.util.*

/**
 * represents virtual [Resource] (means physical existence is not required)
 */
open class VirtualResource<R : VirtualResource<R>> : AbstractResource<VirtualResource<R>> {
    private val baseResource: VirtualResource<R>
    final override lateinit var locale: Locale
        private set

    protected constructor(baseResource: VirtualResource<R>, location: String) : super(location) {
        this.baseResource = baseResource
    }

    protected constructor(location: String) : super(location) {
        @Suppress("LeakingThis")
        baseResource = this
    }

    override fun forLocale(locale: Locale): Resource {
        val control: ResourceBundle.Control =
            ResourceBundle.Control.getControl(ResourceBundle.Control.FORMAT_PROPERTIES)
        val baseName = baseResource.location.substringBeforeLast('.')
        val suffix = baseResource.location.substringAfterLast('.')
        return control.getCandidateLocales(baseName, locale).map {
            createResource(baseResource, control.toResourceName(control.toBundleName(baseName, it), suffix))
        }.firstOrNull { it.exists() }
            ?.let { it.also { it.locale = locale } }
            ?: baseResource
    }

    override fun createResource(base: VirtualResource<R>?, location: String): VirtualResource<R> =
        base?.let { VirtualResource(it, location) } ?: VirtualResource(location)

    override fun toURL(): URL = if (baseResource == this) super.toURL() else URL(baseResource.toURL(), location)

    val baseName: String
        get() = location.substringBeforeLast(".")

    override fun withExtension(extension: String): Resource? =
        createResource(null, "${location.substringBeforeLast('.')}.$extension")

    override val isBase: Boolean
        get() = baseResource === this
    val extension: String
        get() = location.substringAfterLast(".")
    val name: String
        get() {
            return if (!location.contains('/')) {
                if (!location.contains(File.separatorChar)) location
                else location.substringAfterLast(File.separatorChar)
            } else {
                location.substringAfterLast('/')
            }
        }
    val relativeLocation: String
        get() = (0 until location.length.coerceAtMost(baseResource.location.length))
            .firstOrNull { location[it] != baseResource.location[it] }
            ?.let { location.substring(it) }
            ?: location

}
